File: benchmarks/livermore/k16_monte_carlo.c
/* K16 — Monte Carlo Search
   Original Livermore Loop kernel 16.
   Simplified translation of the goto/break/continue zone traversal.
   Arrays: d[300], plan[300], zone[300] (integer).
   n=100, ii=n/3, lb=ii+ii. Scalars: r,s,t,tmp. */
#include <stdio.h>
#include <time.h>
#include <stdint.h>
#include "signel.h"

#define N     100
#define NREPS 100

int main(void) {
    double d[300], plan[300];
    int zone[300];
    int ii = N / 3;
    int lb = ii + ii;
    double spacer[39]; signel(spacer, 39);
    double r = spacer[29], s = spacer[31], t = spacer[35];

    /* Initialise arrays */
    signel(d, 300);
    signel(plan, 300);
    for (int i = 0; i < 300; i++)
        zone[i] = i % (N + 1);
    zone[0] = 5;  /* max m value */

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    int k2, k3, m, i1, j2, j4, j5;
    double tmp;
    for (int l = 0; l < NREPS; l++) {
        i1 = 1; m = 1; k2 = 0; k3 = 0;
        int done = 0;
        while (!done) {
            j2 = (N + N) * (m - 1) + 1;
            int brk = 0;
            for (int k = 1; k <= N && !brk; k++) {
                k2++;
                j4 = j2 + k + k;
                /* Clamp j4 indices to array bounds */
                int j4m1 = (j4 - 1) % 300; if (j4m1 < 0) j4m1 += 300;
                int j4m2 = (j4 - 2) % 300; if (j4m2 < 0) j4m2 += 300;
                j5 = zone[j4m1];
                if (j5 < 0) j5 = -j5;
                if (j5 >= 300) j5 = j5 % 300;

                if (j5 < N) {
                    if (j5 + lb < N)      tmp = plan[j5] - t;
                    else if (j5 + ii < N) tmp = plan[j5] - s;
                    else                   tmp = plan[j5] - r;
                } else if (j5 == N) {
                    brk = 1;
                    continue;
                } else {
                    k3++;
                    int j5m2 = (j5 - 2) % 300; if (j5m2 < 0) j5m2 += 300;
                    int j5m3 = (j5 - 3) % 300; if (j5m3 < 0) j5m3 += 300;
                    int j5m4 = (j5 - 4) % 300; if (j5m4 < 0) j5m4 += 300;
                    int j5m5 = (j5 - 5) % 300; if (j5m5 < 0) j5m5 += 300;
                    int j5m1 = (j5 - 1) % 300; if (j5m1 < 0) j5m1 += 300;
                    tmp = d[j5m1] - (d[j5m2] * (t - d[j5m3]) * (t - d[j5m3])
                          + (s - d[j5m4]) * (s - d[j5m4])
                          + (r - d[j5m5]) * (r - d[j5m5]));
                }

                int zval = zone[j4m2];
                if (tmp < 0.0) {
                    if (zval < 0) { /* continue */ }
                    else if (zval == 0) { brk = 1; continue; }
                    else { /* fall through to m update */ }
                } else if (tmp > 0.0) {
                    if (zval > 0) { /* continue */ }
                    else if (zval == 0) { brk = 1; continue; }
                    else { /* fall through to m update */ }
                } else {
                    brk = 1; continue;
                }

                /* Only reach here when we need to update m and restart */
                if (tmp < 0.0 && zval < 0) continue;
                if (tmp > 0.0 && zval > 0) continue;

                m++;
                if (m > zone[0]) m = 1;
                if (i1 != m) {
                    brk = 1;  /* restart outer while */
                } else {
                    brk = 1;
                }
            }
            if (!brk || i1 == m) done = 1;
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("k2 = %d, k3 = %d\n", k2, k3);
    return 0;
}

File: benchmarks/livermore/k16_monte_carlo.f
C     Kernel 16 -- Monte Carlo search loop
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C
C     This is the original Fortran kernel with GOTO-based control flow.
C     The loop searches zone/plan/d arrays using a Monte Carlo pattern
C     with labeled jumps for break/continue/restart semantics.
C
      PROGRAM K16
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER ZONE
      DIMENSION D(300), PLAN(300), ZONE(300)

      N  = 75
      II = N/3
      LB = II+II

C     Initialize arrays (simplified SIGNEL)
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 10 I = 1, 300
         BUZZ = (1.0D0 - FUZZ)*BUZZ + FUZZ
         FUZZ = -FUZZ
         D(I)    = (BUZZ - FIZZ)*0.1D0
         PLAN(I) = (BUZZ - FIZZ)*0.1D0
   10 CONTINUE
      DO 15 I = 1, 300
         ZONE(I) = MOD(I-1, N+1)
   15 CONTINUE
      ZONE(1) = 5

C     Spacer init for r,s,t (simplified)
      R = 0.1D0
      S = 0.1D0
      T = 0.1D0

      DO 500 IREP = 1, 10000
         I1 = 1
         M  = 1
         K2 = 0
         K3 = 0

C        -- Entry point for restart with new M
  410    J2 = (N+N)*(M-1) + 1
         DO 470 K = 1, N
            K2 = K2 + 1
            J4 = J2 + K + K
            J5 = ZONE(J4-1)
            IF (J5 .LT. N) THEN
               IF (J5+LB .LT. N) THEN
                  TMP = PLAN(J5) - T
               ELSE IF (J5+II .LT. N) THEN
                  TMP = PLAN(J5) - S
               ELSE
                  TMP = PLAN(J5) - R
               END IF
            ELSE IF (J5 .EQ. N) THEN
               GO TO 480
            ELSE
               K3 = K3 + 1
               TMP = D(J5-1) - (D(J5-2)*(T-D(J5-3))*(T-D(J5-3))
     &              +(S-D(J5-4))*(S-D(J5-4))
     &              +(R-D(J5-5))*(R-D(J5-5)))
            END IF

            IF (TMP .LT. 0.0D0) THEN
               IF (ZONE(J4-2) .LT. 0) GO TO 470
               IF (ZONE(J4-2) .EQ. 0) GO TO 480
            ELSE IF (TMP .GT. 0.0D0) THEN
               IF (ZONE(J4-2) .GT. 0) GO TO 470
               IF (ZONE(J4-2) .EQ. 0) GO TO 480
            ELSE
               GO TO 480
            END IF

            M = M + 1
            IF (M .GT. ZONE(1)) M = 1
            IF (I1-M) 410, 480, 410
  470    CONTINUE
  480    CONTINUE
  500 CONTINUE

      WRITE(*,*) 'K16 monte carlo: k2 =', K2, ', k3 =', K3
      END
